### PR TITLE
core renormalization and smarter multiplication by a number

### PR DESCRIPTION
**Pull request by KhrulkovV**

### CODE REVIEW

File: t3f/ops.py [hunk 1]
```diff
@@ -751,7 +751,11 @@ def multiply(tt_left, right):
   if not isinstance(right, TensorTrainBase):
     # Assume right is a number, not TensorTrain.
     tt_cores = list(tt_left.tt_cores)
-    tt_cores[0] = right * tt_cores[0]
+    fact = tf.pow(tf.cast(tf.abs(right), tt_left.dtype), 1.0/(tt_left.ndims()))
```
Comment (Bihaqo) [inline t3f/ops.py]: I'm not sure if it's a good idea to slow down the multiplication.. Maybe add a flag?

Comment (Bihaqo) [reply t3f/ops.py]: Also please add spaces around '/' and add a comment about that's happening here

File: t3f/ops.py [hunk 2]
```diff
@@ -1132,3 +1136,50 @@ def gather_nd(tt, indices):
     tt_elements = tf.matmul(tt_elements, core_slices)
   tt_elements = tf.reshape(tt_elements, tf.shape(indices)[:-1])
   return tt_elements
+
+def renormalize_cores(tt):
```
Comment (Bihaqo) [inline t3f/ops.py]: maybe better renormalize_tt_cores? I tried not to say the word 'cores' (and use tt-cores instead) anywhere but in the internal code.

File: t3f/ops.py [hunk 3]
```diff
@@ -1132,3 +1136,50 @@ def gather_nd(tt, indices):
...
+    """Renormalizes tt cores to make them of the same Frobenius norm.
+
+    Does not change tt, but makes computations with cores more stable.
```
Comment (Bihaqo) [inline t3f/ops.py]: Doesn't change the tensor represented by the `tt` object, but renormalizes the TT-cores to make further computations more numerically stable.

File: t3f/ops.py [hunk 4]
```diff
@@ -1132,3 +1136,50 @@ def gather_nd(tt, indices):
...
+      tensor as tt, but with all cores having equal norm. In the batch
+      case applies to each TT in `TensorTrainBatch`.
+
+    """
+    if isinstance(tt, TensorTrain):
+      new_cores = []
+      running_log_norm = 0
+      core_norms = []
+      for core in tt.tt_cores:
+        cur_core_norm = tf.sqrt(tf.reduce_sum(core ** 2) + 1e-8)
```
Comment (Bihaqo) [inline t3f/ops.py]: Why do you need this eps, because of the gradient? Please, add it to the arguments (with 1e-8 being default value), make sure that this default value works with tf.float32, and think if we can avoid having eps whatsoever :)

File: t3f/ops.py [hunk 5]
```diff
@@ -1132,3 +1136,50 @@ def gather_nd(tt, indices):
...
+        core_norms.append(cur_core_norm)
+        running_log_norm += tf.log(cur_core_norm)
+
+      running_log_norm = running_log_norm / tt.ndims()
+      fact = tf.exp(running_log_norm)
+      for i, core in enumerate(tt.tt_cores):
+        new_cores.append(core * fact / core_norms[i])
```
Comment (Bihaqo) [inline t3f/ops.py]: Why do you need to divide by core_norms?

File: t3f/ops_test.py [hunk 6]
```diff
@@ -182,7 +196,7 @@ def testFullMatrix2d(self):
       with self.test_session():
         tf_mat = TensorTrain(tt_cores)
         actual = ops.full(tf_mat)
-        self.assertAllClose(desired, actual.eval())
+        self.assertAllClose(desired, actual.eval(), atol=1e-5, rtol=1e-5)
```
Comment (Bihaqo) [inline t3f/ops_test.py]: Why did you have to change this?

File: t3f/ops_test.py [hunk 7]
```diff
@@ -693,6 +707,20 @@ def testGatherNDBatch(self):
       self.assertAllClose(res_np_v, des_v)
       self.assertAllClose(res_pl_v, res_pl_v)
 
+  def testCoreRenormBatch(self):
+      a = initializers.random_tensor_batch(3 * (10,), tt_rank=7, batch_size=5)
```
Comment (Bihaqo) [inline t3f/ops_test.py]: This class is for testing matrices, not tensors

File: t3f/ops.py (hunk 5 above)
Comment (KhrulkovV) [reply t3f/ops.py]: To make each core of the norm "fact", which is equal to (||C_1|| ||C_2|| ... ||C_d||)^(1/d). So I divide each core by its norm, and multiply by "fact". This doesn't change overall tensor.

Comment (Bihaqo) [reply t3f/ops.py]: Make sense, sorry for being stupid :)

Comment (Bihaqo) [reply t3f/ops.py]: I don't see how to close this thread. If you don't either, just ignore it)

File: t3f/ops_test.py (hunk 7 above)
Comment (KhrulkovV) [reply t3f/ops_test.py]: But this function was in `TTTensorBatchTest` class ...

File: t3f/ops.py [hunk 8]
```diff
@@ -747,16 +750,28 @@ def multiply(tt_left, right):
   is_right_batch = isinstance(right, TensorTrainBatch)
 
   is_batch_case = is_left_batch or is_right_batch
-
+  ndims = tt_left.ndims()
   if not isinstance(right, TensorTrainBase):
     # Assume right is a number, not TensorTrain.
     tt_cores = list(tt_left.tt_cores)
-    tt_cores[0] = right * tt_cores[0]
+    if not fast:
```
Comment (Bihaqo) [inline t3f/ops.py]: Sorry, after all lets not do this flag and just always be robust and a bit slow. Can you please revert the flag part?

File: t3f/ops.py [hunk 9]
```diff
@@ -1132,3 +1147,51 @@ def gather_nd(tt, indices):
     tt_elements = tf.matmul(tt_elements, core_slices)
   tt_elements = tf.reshape(tt_elements, tf.shape(indices)[:-1])
   return tt_elements
+
+def renormalize_tt_cores(tt, epsilon=1e-8):
```
Comment (Bihaqo) [inline t3f/ops.py]: (pep8) Please, use two blank lines before high-level function definition (note that it does NOT apply to class methods).
